cmd/ido2icingadb/misc.go

```diff
@@ -0,0 +1,387 @@
+package main
+
+import (
+	"context"
+	"crypto/sha1"
+	"fmt"
+	"github.com/icinga/icingadb/pkg/contracts"
+	"github.com/icinga/icingadb/pkg/driver"
+	"github.com/icinga/icingadb/pkg/icingadb"
+	"github.com/icinga/icingadb/pkg/icingadb/objectpacker"
+	icingadbTypes "github.com/icinga/icingadb/pkg/types"
+	"github.com/jmoiron/sqlx"
+	"github.com/pkg/errors"
+	"github.com/vbauerster/mpb/v6"
+	"github.com/vbauerster/mpb/v6/decor"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"strings"
+	"time"
+)
+
+// slowlyIncrementingProgressDecorator is the base decorator
+// for possibly slowly incrementing progresses possibly starting from >0%.
+type slowlyIncrementingProgressDecorator struct {
+	decor.WC
+
+	// startProgress is the first progress >0 seen by Decor.
+	startProgress int64
+	// startTime tells when is startProgress from.
+	startTime time.Time
+	// lastProgress is the last progress >0 seen by Decor.
+	lastProgress int64
+	// lastTime tells when is lastProgress from.
+	lastTime time.Time
+}
+
+// update is to be called by Decor and returns whether sipd is warmed up.
+func (sipd *slowlyIncrementingProgressDecorator) update(s decor.Statistics) bool {
+	if s.Completed || s.Current < 1 {
+		return false
+	}
+
+	if sipd.startProgress < 1 {
+		sipd.startProgress = s.Current
+		sipd.startTime = time.Now()
+		sipd.lastProgress = sipd.startProgress
+		sipd.lastTime = sipd.startTime
+
+		return false
+	}
+
+	if s.Current == sipd.startProgress {
+		return false
+	}
+
+	if s.Current > sipd.lastProgress {
+		sipd.lastProgress = s.Current
+		sipd.lastTime = time.Now()
+	}
+
+	return true
+}
+
+// eta indicates the ETA for possibly slowly incrementing progresses possibly starting from >0%.
+type eta struct {
+	slowlyIncrementingProgressDecorator
+}
+
+// Decor implements the decor.Decorator interface.
+func (e *eta) Decor(s decor.Statistics) string {
+	if e.update(s) {
+		timePerItem := float64(e.lastTime.Sub(e.startTime)) / float64(e.lastProgress-e.startProgress)
+		lastETA := time.Duration(float64(s.Total-s.Current) * timePerItem)
+
+		return e.FormatMsg(((lastETA - time.Since(e.lastTime)) / time.Second * time.Second).String())
+	} else {
+		return ""
+	}
+}
+
+// eta indicates ops/s for possibly slowly incrementing progresses possibly starting from >0%.
+type opsPerSec struct {
+	slowlyIncrementingProgressDecorator
+}
+
+// Decor implements the decor.Decorator interface.
+func (ops *opsPerSec) Decor(s decor.Statistics) string {
+	if ops.update(s) {
+		return ops.FormatMsg(fmt.Sprintf(
+			"%.0f/s",
+			float64(ops.lastProgress-ops.startProgress)/
+				(float64(ops.lastTime.Sub(ops.startTime))/float64(time.Second)),
+		))
+	} else {
+		return ""
+	}
+}
```

Comment:
The progress bars use the average from start to show rates and time estimates. Given that there are index updates involved when inserting, insert complexity should be (at least) logarithmic, i.e. how the insert rate progresses over time can be approximated by [1/ln(x)](https://www.wolframalpha.com/input?i=plot+1%2Fln%28x%29+for+x+from+1+to+1e6). So inserting will likely only get slower over time (unless there's a magic performance improvement on the database server in the meantime), making the values displayed too optimistic.

Thus, I think displaying an average value over the last few minutes or transactions (to avoid having the exact numbers jump around like crazy) would produce more useful numbers here.

Reply:
Is it worth investing the time?

Reply:
Shouldn't take that much time I think. Should just need a slice of progress/time pairs where you keep the last like 10 updates and then perform the same calculation just with the oldest and newest values from that slice.

Reply:
Feel free to continue reviewing while I'm on it.